relays/lib-substrate-relay/src/messages_target.rs

```diff
@@ -258,27 +258,88 @@ where
 					P::SourceChain::NAME,
 				))
 			})?;
+
+		// Prepare 'dummy' delivery transaction - we only care about its length and dispatch weight.
+		let delivery_tx = self.lane.make_messages_delivery_transaction(
+			Zero::zero(),
+			HeaderId(Default::default(), Default::default()),
+			nonces.clone(),
+			prepare_dummy_messages_proof::<P::SourceChain>(nonces.clone(), total_dispatch_weight, total_size),
+		);
+		let delivery_tx_fee = self.client.estimate_extrinsic_fee(delivery_tx).await?;
+		let inclusion_fee_in_target_tokens = delivery_tx_fee.inclusion_fee();
+
+		// The pre-dispatch cost of delivery transaction includes additional fee to cover dispatch fee payment
+		// (Currency::transfer in regular deployment). But if message dispatch has already been paid
+		// at the Source chain, the delivery transaction will refund relayer with this additional cost.
+		// But `estimate_extrinsic_fee` obviously just returns pre-dispatch cost of the transaction. So
+		// if transaction delivers prepaid message, then it may happen that pre-dispatch cost is larger
+		// than reward and `Rational` relayer will refuse to deliver this message.
+		//
+		// The most ovious solution would be to deduct total weight of dispatch fee payments from the
+		// `total_dispatch_weight` and use regular `estimate_extrinsic_fee` call. But what if
+		// `total_dispatch_weight` is less than total dispatch fee payments weight? Weight is strictly
+		// positive, so we can't use this option.
+		//
+		// Instead we'll be directly using `WeightToFee` and `NextFeeMultiplier` of the Target chain.
+		// This requires more knowledge of the Target chain, but seems there's no better way to solve
+		// this now.
+		let expected_refund_in_target_tokens = if total_prepaid_nonces != 0 {
+			const WEIGHT_DIFFERENCE: Weight = 100;
+
+			let larger_dispatch_weight = total_dispatch_weight.saturating_add(WEIGHT_DIFFERENCE);
+			let larger_delivery_tx_fee = self
+				.client
+				.estimate_extrinsic_fee(self.lane.make_messages_delivery_transaction(
+					Zero::zero(),
+					HeaderId(Default::default(), Default::default()),
+					nonces.clone(),
+					prepare_dummy_messages_proof::<P::SourceChain>(nonces.clone(), larger_dispatch_weight, total_size),
+				))
+				.await?;
+
+			compute_prepaid_messages_refund::<P>(
+				total_prepaid_nonces,
```

Comment:
Do I get this right, that we take into account cumulative reward from all pre-paid transactions here? Can you elaborate a little bit more on this? For instance a more concrete example would help.

Reply:
Delivery transaction contains some messages (a number of elements in the `nonces` range). The cost of dispatch of some of these messages is already paid `AtSourceChain` (number of such messages is what `total_prepaid_nonces` means). Dispatch of other messages is paid `AtTargetChain`. For the former group, delivery tx will deduct a weight of `Currency::transfer` call (payment for dispatch) from its pre-dispatch weight. This sum (we know the weight, but need to compute cost) is computed here (`expected_refund_in_target_tokens`).

Then (a bit above) there's also a regular call to the `estimate_extrinsic_fee`, which just estimates a cost of delivery transaction without taking into account any refunds (result is `inclusion_fee_in_target_tokens`). So final cost will be `inclusion_fee_in_target_tokens - expected_refund_in_target_tokens`.